keps/sig-storage/2317-fsgroup-on-mount/README.md

```diff
@@ -0,0 +1,451 @@
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Currently for most volume plugins kubelet applies fsgroup ownership and permission based changes by recursively `chown`ing and `chmod`ing the files and directories inside a volume. For certain CSI drivers this may not be possible because `chown` and `chmod` are unix primitives and underlying CSI driver
+may not support them. This enhancement proposes providing the CSI driver with fsgroup as an explicit field so as CSI driver can apply this on mount time.
+
+## Motivation
+
+Since some CSI drivers(AzureFile for example) - don't support chmod/chown - we propose that fsgroup of pod to be provided to the CSI driver on `NodeStage`
+and `NodePublish` CSI RPC calls. This will allow the CSI driver to apply `fsgroup` as a mount option on `NodeStage` or `NodePublish` and kubelet can be freed
+from responsibility of applying recursive ownership and permission change.
+
+### Goals
+
+- Allow CSI driver to mount volumes with provided fsgroup.
+
+### Non-Goals
+
+- We are not supplying `fsGroup` as a generic ownerhip and permission handle to the CSI driver. We do not expect CSI drivers to `chown` or `chmod` files.
+
+## Proposal
+
+We are updating CSI specs by adding additional field called `volume_mount_group` to `NodeStage` and `NodePublish` RPC calls. The CSI proposal is available at - https://github.com/container-storage-interface/spec/pull/468 .
+
+The CSI spec change is deliberately trying to avoid asking drivers to use supplied `fsGroup` as a generic handle for ownership and permissions. The reason being - Kubernetes may expect ownership and permissions to be in a way that is very platform/OS specific. We do not think CSI driver is right place to enforce all kind of different permissions expected by Kubernetes. The full scope of that discussion is out of scope for this enhancement and interested folks can follow along on - https://github.com/container-storage-interface/spec/issues/449
+
+
+### Risks and Mitigations
+
+I am not aware of any associated risks. If a driver can not support using `fsgroup` as a mount option, it can always use `FileFSGroupPolicy` and let kubelet handle the ownership and permissions.
```

Comment:
Please elaborate somewhere that there is a behavioral change: if two pods use the same volume with different `fsGroup`, then:
* Before this change, the pod that starts as second will `chown` all files, so the first pod probably can't access them.
* After this change, both pods may access the volume, if the volume can be mounted twice with different options.

This change is neligible and most probably corresponds to what in-tree Azure File plugin already does today.

Reply:
No before this change - no chown/chmod will be performed at all and hence except the root user(or pods that run as root),  no pod will be able to read/write to the volume. 

I am not sure if this is a behaviour change. Before - only root could read/write to the pod and now other users can too.  I could add something here but I am not sure if it fits under risks and mitigations.

Reply:
A CSI driver could have supported `chown` / `chmod` for RWO / RWX volumes (such as github.com/kubernetes-csi/csi-driver-smb, if the samba export is configured properly), it's just faster with a mount option. Users of this driver could experience a very small behavior change.

Reply:
addressed under the section - `Does enabling the feature change any default behavior?**`